SimTracker/Common/interface/SiPixelChargeReweightingAlgorithm.h

```diff
@@ -841,4 +848,219 @@ bool SiPixelChargeReweightingAlgorithm::lateSignalReweight(const PixelGeomDetUni
+  track[4] = direction.y();
+  track[5] = direction.z();
+
+  array_2d pixrewgt(boost::extents[TXSIZE][TYSIZE]);
+
+  /*
+  for (int row = 0; row < TXSIZE; ++row) {
+    for (int col = 0; col < TYSIZE; ++col) {
+      pixrewgt[row][col] = 0;
+    }
+  }
+*/
+
+  for (int row = 0; row < TXSIZE; ++row) {
+    xdouble[row] = topol->isItBigPixelInX(hitPixel.first + row - THX);
+  }
+
+  for (int col = 0; col < TYSIZE; ++col) {
+    ydouble[col] = topol->isItBigPixelInY(hitPixel.second + col - THY);
+  }
+
+  for (int row = 0; row < TXSIZE; ++row) {
+    for (int col = 0; col < TYSIZE; ++col) {
+      //Fill charges into 21x13 Pixel Array with hitPixel in centre
+      pixrewgt[row][col] =
+          theDigiSignal[PixelDigi::pixelToChannel(hitPixel.first + row - THX, hitPixel.second + col - THY)];
+    }
+  }
+
+  if (PrintClusters) {
+    LogDebug("SiPixelChargeReweightingAlgorithm") << " Cluster before reweighting: ";
+    printCluster(pixrewgt);
+  }
+
+  int ierr;
+  // for unirradiated: 2nd argument is IDden
+  // for irradiated: 2nd argument is IDnum
+  int ID1 = dbobject_num->getTemplateID(detID);
+  int ID0 = dbobject_den->getTemplateID(detID);
+
+  if (ID0 == ID1) {
+    LogDebug("SiPixelChargeReweightingAlgorithm") << " same template for num and den ";
+    return false;
+  }
+  ierr = PixelTempRewgt2D(ID0, ID1, pixrewgt);
+  if (ierr != 0) {
+    LogDebug("SiPixelChargeReweightingAlgorithm") << "Cluster Charge Reweighting did not work properly.";
+    return false;
+  }
+  if (PrintClusters) {
+    LogDebug("SiPixelChargeReweightingAlgorithm") << " Cluster after reweighting: ";
+    printCluster(pixrewgt);
+  }
+
+  for (int row = 0; row < TXSIZE; ++row) {
+    for (int col = 0; col < TYSIZE; ++col) {
+      float charge = 0;
+      charge = pixrewgt[row][col];
+      if ((hitPixel.first + row - THX) >= 0 && (hitPixel.first + row - THX) < topol->nrows() &&
+          (hitPixel.second + col - THY) >= 0 && (hitPixel.second + col - THY) < topol->ncolumns() && charge > 0) {
+        chargeAfter += charge;
+        if constexpr (std::is_same_v<AmplitudeType, digitizerUtility::Ph2Amplitude>) {
+          edm::LogError("SiPixelChargeReweightingAlgorithm")
+              << "Phase-2 late charge reweighting not supported (not sure we need it at all)";
+          return false;
+        } else {
+          theNewDigiSignal[PixelDigi::pixelToChannel(hitPixel.first + row - THX, hitPixel.second + col - THY)] +=
+              AmplitudeType(charge, charge);
+        }
+      }
+    }
+  }
+
+  if (chargeBefore != 0. && chargeAfter == 0.) {
+    return false;
+  }
+
+  if (PrintClusters) {
+    LogDebug("SiPixelChargeReweightingAlgorithm")
+        << "Charges (before->after): " << chargeBefore << " -> " << chargeAfter;
+    LogDebug("SiPixelChargeReweightingAlgorithm") << "Charge loss: " << (1 - chargeAfter / chargeBefore) * 100 << " %";
+  }
+
+  // need to store the digi out of the 21x13 box.
+  for (auto& i : theDigiSignal) {
```

Comment:
@ferencek , may be "auto const&" ?

Reply:
Addressed here https://github.com/cms-sw/cmssw/pull/46884/commits/fd11b453cbff4efc8602bd61ae604a5461d654b0